### ISSUE TITLE
[UX] Hide "Run Database Updates" / next steps if there are no next steps

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#3245]
## Describe your issue or idea
When updating a module via the UI, there is not always a database update to complete, but the "Next Steps" section always shows with the "Run database updates" link:

![screen shot 2018-08-07 at 10 52 11 am](https://user-images.githubusercontent.com/871421/43787493-bcad06ee-9a30-11e8-8eb6-74d1ee4bb72a.jpg)

When you click the link, you are informed that there are no database updates pending (and a warning about backing up before you run updates). 

![screen shot 2018-08-07 at 10 51 54 am](https://user-images.githubusercontent.com/871421/43787523-ce36f7bc-9a30-11e8-9c81-2985e9ab8609.jpg)

It would be better to just display a "No updates pending" message after the first step.

**Comment by klonos** [comment]
Yeah, thanks for filing this ticket @laryn 👍 ...I don't understand what is the use of the "project_name: Installed project_name successfully" list.

I also do not understand why we are listing "Your projects have been downloaded and updated." as a task. It feels like it should be a message, but we already have "Update was completed successfully." in place, so perhaps we should just remove it altogether.

If we remove the "Your projects have been downloaded and updated." task, then we are left only with the "Run database updates" task, in which case, why aren't we detecting whether there are any db updates or not and act accordingly (either redirect to `/core/update.php`, or back to `/admin/modules/update`)?

Also noting that this step is in `/authorize.php`, which we are looking to remove (#3208).

**Comment by klonos** [comment]
...yep, the code for this is in [`installer_authorize_update_batch_finished()`](https://github.com/backdrop/backdrop/blob/1.x/core/modules/installer/installer.authorize.inc#L185), which states:

>Batch callback: Performs actions when the authorized update batch is done.
> 
> This processes the results and stashes them into SESSION such that
> authorize.php will render a report. Also responsible for putting the site
> back online and clearing the update status cache after a successful update.

**Comment by klonos** [comment]
...when installing a module (as opposed to when updating), you get this screen instead:

![enablemodule](https://cloud.githubusercontent.com/assets/14132746/10231291/8f646ede-6847-11e5-9351-52daab397277.png)

The list of tasks here makes more sense.

**Comment by laryn** [comment]
Just ran into this again while testing core update via the UI. At a minimum it would be a cleaner experience without the "Run database updates" link under "Next steps".

**Comment by yorkshire-pudding** [comment]
I totally agree with this ticket, we need to focus on what they need to do.  Very related to this workflow is how the UI handles maintenance mode for installs and updates.
It advises you to put into maintenance mode to install the module/update and gives you a handy tick box to do it.  Then it takes you out of maintenance mode.  It then (let's assume there are database updates) says to do the database updates and recommends they're done in maintenance mode, however there is no option to re-enable it; you would need to come out re-enable maintenance mode then go back and activate the database updates.

What it should do, taking into account the above:

1. Download modules/updates
2. Recommend to put into maintenance mode and keep the option
3. If there are database updates, keep it in maintenance mode
4. Once database updates are done take out of maintenance mode

Let me know if this needs to be a separate issue.

**Comment by stpaultim** [comment]
What is the easiest way to test this?

**Comment by laryn** [comment]
@stpaultim 
Install an outdated version of a module in the sandbox (or locally on a site running this patch), making sure that there are no database updates in between the version you install and the latest version. Update the module through the UI and try to verify that it doesn't walk you through to the database updates page, and that the module does actually get successfully updated.

**Comment by klonos** [comment]
I'll have a look at the PR soon, I promise.

> Just ran into this again while testing core update via the UI.

Yeah, I just run the recent security update for core and got this:

![image](https://user-images.githubusercontent.com/2423362/158710719-66f264f4-2de3-4fc4-9366-ed9ac9b26da9.png)

I would suggest a few improvements:
- split the 2 different things (update finished successfully / out of maintenance mode) into separate messages
- use the human name of the module(s)/project(s) being updated
- list multiple projects using `them_items_list()`

**Comment by yorkshire-pudding** [comment]
I'd like to reiterate [my suggestion](https://github.com/backdrop/backdrop-issues/issues/3245#issuecomment-985324465) that the process should handle maintenance mode for the database updates. If we recommend (and I agree with the recommendation) that they put the site into maintenance mode for the database updates, then we should help them with that.

**Comment by laryn** [comment]
@klonos I'm still interested in this but won't have time to look at it in the near future. Feel free to take over if you wish!

**Comment by herbdool** [comment]
@yorkshire-pudding I've marked this needing testing since it's not clear if you tested the two scenarios (see my other comment on the PR).

By the way, not sure if this could be part of this issue, but wondering if it would be better to rename "Run database updates" to "Run database and configuration updates" since the same process is used for making config updates. It might be required to update this elsewhere in the code too so could be a new issue.

**Comment by yorkshire-pudding** [comment]
Doesn't prompt for database updates when there are any.  Tested with first release of `paragraphs` manually installed via command line (to avoid the issue in #3997) and then doing an update.  On another site (and going to database updates via status report after) there are 4 updates that should have been done.

**Comment by argiepiano** [comment]
I've verified that this patch fails to detect needed updates and therefore fails to show the `Run database updates` links. I left a comment in the PR after @yorkshire-pudding's one.

**Comment by argiepiano** [comment]
Alternative PR: https://github.com/backdrop/backdrop/pull/4060

(@laryn said he's OK with alternative ones since he is away this weekend)

**Comment by stpaultim** [comment]
In my initial tests, this one does not seem to be working for me. The module is not getting updated. After I run the update, it takes me back here and it appears the module has not been updated.

![image](https://user-images.githubusercontent.com/3144571/166133727-faad6726-3e6f-4b7a-870a-ad719682cfd7.png)

I'm going to test again to be sure I didn't do something wrong.

**Comment by stpaultim** [comment]
Sorry @argiepiano, but your PR is not working for me. I created a new local version of the site using this:
https://github.com/argiepiano/backdrop/tree/1.x-issue-3245-argiepiano

I then experimented with the "simplify" module. I initially installed:
https://github.com/backdrop-contrib/simplify/releases/tag/1.x-1.2.0

I then try to run the update process, but from this page:
![image](https://user-images.githubusercontent.com/3144571/166134206-9efd1b80-df94-4126-bca6-8ad952c12080.png)

It takes me back to this page:
![image](https://user-images.githubusercontent.com/3144571/166134216-73d8cfeb-d99a-4983-ac91-155ba03c55a3.png)

And the update has not been done.

**Comment by argiepiano** [comment]
That's strange for two reasons:
- I followed your steps and it's working for me - which leads me to think that perhaps there is something missing in your test or on its description
- My PR is basically the same as @laryn's with one small difference. The PR does not interfere at all with the update process.

Mind testing again with a clean install?

**Comment by yorkshire-pudding** [comment]
@argiepiano - I've tested with a clean install and the PR applied as a patch.  I've verified that the patch is applied.  Installed entity_plus (dependency of paragraphs).  Installed old version of paragraphs.
Ran update.  Downloaded updates but didn't prompt for database updates.  Went to `core/update.php` and there were 4 database updates pending.

![chrome_ZrmJsoN7aT](https://user-images.githubusercontent.com/20357797/166162705-2b23c5b7-0c90-4866-a955-df4c2a401d34.gif)

**Comment by argiepiano** [comment]
Thanks for testing and for the details @yorkshire-pudding. So it looks like my PR did not solve this issue. I tested with a different module, but I will test with Paragraphs to try to find out what's going on.

**Comment by argiepiano** [comment]
@yorkshire-pudding I've found the issue - Paragraphs 1.0.0 had a schema of >7000, so the check was failing. I've made a new commit. Please test. Thanks

**Comment by yorkshire-pudding** [comment]
@argiepiano - that works for with db update.  Will quickly find a suitable case for testing without to cover all bases.

![chrome_miADXOBf9F](https://user-images.githubusercontent.com/20357797/166166574-6920d515-42cf-4777-a7f0-17296399939c.gif)

**Comment by argiepiano** [comment]
You can use Entity Plus version 15 and version 16

**Comment by argiepiano** [comment]
I just realized that there are cases where a submodule may have updates. My PR only loads the install files for the main modules. I think I need to revert to loading the install files for all the enabled modules. I'm going to modify the PR and commit those changes in a minute

**Comment by yorkshire-pudding** [comment]
Ok.  Tested with Site_alert and all looked good
![chrome_YjjLw1kBtw](https://user-images.githubusercontent.com/20357797/166166807-2477d6fe-36de-456e-bc2d-61b6570ec660.gif)
d

**Comment by argiepiano** [comment]
This is going to be trickier than it seemed at first.

Even the current PR, which worked for you, will fail if there are more than one module being updated. The issue is that `backdrop_get_schema_versions()` static-caches the list of enabled modules and uses that cache to check for hook_update_N in them in the future. The problem is that the call to `backdrop_get_schema_versions()` inside the new code in `installer.authorize.inc` happens at a point where we are apparently in bootstrap or maintenance mode, not in runtime mode, and therefore `module_list()` inside `backdrop_get_schema_versions()` only return 2 modules: system and update, which are cached. So the next time `backdrop_get_schema_versions()` is invoked (when there are more than one module being updated), that limited cache will be used, and therefore the new code won't detect that there are updates available.

So... this is getting complicated. One solution may be to to a `backdrop_static_reset('backdrop_get_schema_versions)` in the new code in every iteration before calling `backdrop_get_schema_versions`. Which seems really overkill for a feature that is really not that important (not showing a link to database update, which only admins see, don't seem so crucial).

Perhaps I'll let @laryn decide whether he wants to pick this up and find another way?

**Comment by argiepiano** [comment]
Well, after further testing it turns out that the PR as it stands now DOES work when there are multiple module updates. My PR only iterates on the successfully updated modules to see if there are any database updates to be run. The original PR from @laryn used `module_list()`, which doesn't work in that context. 

However, the issue remains to be tested on whether database updates in submodules (rather than on the main module) will be picked up.

**Comment by klonos** [comment]
These PRs will need to be updated after the change from "database updates" to "site updates" landed with #5630.

@argiepiano, @laryn, do you have capacity to work on this, or should I take a crack at it? (starting off of your PRs)

I also need to check if there's overlap with #512 (or how much overlap)

**Comment by laryn** [comment]
@klonos You are welcome to it. I'm not sure when I'd be able to come back to this one.

**Comment by klonos** [comment]
Thanks @laryn 🙏🏼 ...I believe that we can close your PR. Can you please do that?

@argiepiano I believe that your PR is close. Can you please resolve the conflicts, and have a look at the comments I've left?

**Comment by argiepiano** [comment]
@klonos, I'm on the road for the next 2 weeks, with very little capacity. Feel free to pick up. I'm closing mine.
